### PR TITLE
Added onEnd callback into AnimatedSize

### LINKED ISSUE #106439: Add VoidCallback to AnimatedSize (Similar to AnimatedOpacity's onEnd)
## Use case
I am using an implicit Animation widget and an explicit animating widget, and I would like to be able to add a listener to see when the implicit widget is done to then call my explicitly animated widget. In short,  I would like the AnimatedSize to finish animating before changing the Opacity. And, not stopping there, I would like to be able to chain the animations for all implicitly animated widgets. Something along the lines of 

```dart

AnimatedSize(
    onFinishAnimating: () { 
         driveExplicitAnimation();
         // or 
       
       setState(() { 
           setAnotherAnimatedWidgetsValue();
      });
    }
    child: Container()
);

```

## Proposal

add callback/listener on implicitly animated widget to call explicitly animated widget when the implicitly animated widget is done animating. 

### Note: While the sample code does not have any explicitly animated widgets, it provides the two types of "desired" animations from the implicitly animated widgets. But, because there is nothing that allows me to snoop on their animations (diving into the code for Animated Size, it seems that the animation controller is private, not allowing me to user key.currentState.findRenderObject), thus is not allowing me to better chain animations. When I set the child of the expanded content, I would like the AnimatedSize to finish animating before changing the Opacity. 
<details>

```dart

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
 
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<int> rows = List.filled(5, 0);

  @override
  Widget build(BuildContext context) {

    return Scaffold(
        appBar: AppBar(

          title: Text(widget.title),
        ),
        body: ListView.builder(itemBuilder: ((context, index) {
          return AnimatingRow(rowNum: index, info: "Info at a glance");
        })));
  }
}

class AnimatingRow extends StatefulWidget {
  final int rowNum;
  final String info;

  const AnimatingRow({super.key, required this.rowNum, required this.info});

  @override
  AnimatingRowState createState() => AnimatingRowState();
}

class AnimatingRowState extends State<AnimatingRow> {
  static const String _showMoreText = "Show More";
  static const String _showLessText = "Show Less";
  static const rowColor = Color.fromARGB(255, 235, 235, 235);
  static const fillerText =
      "asdfjaslkdfjlsdjkfalskdjfalasodifj;lasdfjlkasjfd;lkajsdfljkas;dflkjas;dlkfja;sldjfasldkjflasdjf;lsakjdfl;asjdflsdkjfal;sdjf;alsdjfk;lasdjf;lasjdf;lasjdfl;askjdfl;ajsdfjkasdf";
  late bool showingMore;
  late double childOpacity;
  late final int rowNum;
  late String _showText;

  Widget moreWidget = Container();
  @override
  void initState() {
    _showText = _showMoreText;
    showingMore = false;
    childOpacity = 0;
    rowNum = widget.rowNum;
    super.initState();
  }

  Color? getRowColor(int rowNumber) {
    if (rowNumber.isEven) {
      return rowColor;
    } else {
      return null;
    }
  }

  void showButtonPress() {
    if (showingMore) {
      setState(() {
        _showText = _showMoreText;
        showingMore = false;
        moreWidget = Container();
        childOpacity = 0;
      });
    } else {
      const exampleList = Text(fillerText);
      setState(() {
        _showText = _showLessText;
        showingMore = true;
        moreWidget = exampleList;
        childOpacity = 1;
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedSize(
        duration: const Duration(milliseconds: 250),
        child: ColorBackground(
            backgroundColor: getRowColor(rowNum),
            margin: const EdgeInsets.only(
              left: 20,
              right: 20,
            ),
            child: Column(children: [
              Row(
                mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                textDirection: TextDirection.ltr,
                children: [
                  Text(widget.info),
                  Align(
                      alignment: Alignment.bottomRight,
                      child: TextButton(
                        onPressed: showButtonPress,
                        child: Text(_showText),
                      ))
                ],
              ),
              AnimatedOpacity(
                duration: const Duration(milliseconds: 250),
                opacity: childOpacity,
                child: moreWidget,
              )
            ])));
  }
}

class ColorBackground extends Container {
  final double radius;
  final double opacity;
  final MainAxisAlignment mainAxisAlignment;
  late final Color? backgroundColor;

  ColorBackground(
      {super.key,
      super.child,
      this.opacity = 1,
      super.width,
      super.height,
      super.constraints,
      this.backgroundColor = Colors.grey,
      this.radius = 10,
      this.mainAxisAlignment = MainAxisAlignment.center,
      super.padding = const EdgeInsets.all(15),
      super.margin});

  @override
  BoxDecoration get decoration {
    return BoxDecoration(
        color: backgroundColor?.withOpacity(opacity),
        borderRadius: BorderRadius.all(Radius.circular(radius)));
  }
}


```

</details>

### PR DESCRIPTION
**Pull request by feduke-nukem**
close #106439 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/animated_size.dart [hunk 1]
```diff
@@ -363,6 +380,12 @@ class RenderAnimatedSize extends RenderAligningShiftedBox {
     }
   }
 
+  void _animationStatusListener(AnimationStatus status) {
+    if (status == AnimationStatus.completed) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/animated_size.dart:384]: This should be a switch per our style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-if-chains-or--or--with-enum-values

File: packages/flutter/lib/src/widgets/animated_size.dart [hunk 2]
```diff
@@ -78,12 +79,18 @@ class AnimatedSize extends StatefulWidget {
...
+  /// at the end of the current animation.
+  final VoidCallback? onEnd;
+
   @override
   State<AnimatedSize> createState() => _AnimatedSizeState();
 }
 
 class _AnimatedSizeState
-    extends State<AnimatedSize> with SingleTickerProviderStateMixin {
+ extends State<AnimatedSize> with SingleTickerProviderStateMixin {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/animated_size.dart:93]: please undo this whitespace change

File: packages/flutter/test/widgets/animated_size_test.dart [hunk 3]
```diff
@@ -88,6 +102,57 @@ void main() {
...
+            duration: const Duration(milliseconds: 200),
+            child: const SizedBox(
+              width: 200.0,
+              height: 200.0,
+            ),
+          ),
+        ),
+      );
+
+
```
Comment (goderbauer) [inline packages/flutter/test/widgets/animated_size_test.dart:134]: nit: remove one of the blank lines

File: packages/flutter/test/widgets/animated_size_test.dart [hunk 4]
```diff
@@ -88,6 +102,57 @@ void main() {
       expect(box.size.height, equals(100.0));
     });
 
+    testWidgetsWithLeakTracking('calls onEnd when animation is completed', (WidgetTester tester) async {
+      await tester.pumpWidget(
```
Comment (goderbauer) [inline packages/flutter/test/widgets/animated_size_test.dart:106]: You can simplify this. remove the `mockOnEndFunction` stuff and just define an inline function here:

```dart
int callCount = 0;
void handleEnd() {
  callCount++;
}
```

and then just pass `handleEnd` wherever you need it

File: packages/flutter/lib/src/rendering/animated_size.dart (hunk 1 above)
Comment (feduke-nukem) [reply packages/flutter/lib/src/rendering/animated_size.dart]: Done

File: packages/flutter/lib/src/widgets/animated_size.dart (hunk 2 above)
Comment (feduke-nukem) [reply packages/flutter/lib/src/widgets/animated_size.dart]: Done

File: packages/flutter/test/widgets/animated_size_test.dart (hunk 3 above)
Comment (feduke-nukem) [reply packages/flutter/test/widgets/animated_size_test.dart]: Done

File: packages/flutter/test/widgets/animated_size_test.dart (hunk 4 above)
Comment (feduke-nukem) [reply packages/flutter/test/widgets/animated_size_test.dart:97]: Done

**Review by goderbauer — APPROVED** [review]
LGTM

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Pull request merged**